target_chains/sui/contracts/sources/merkle_tree.move

```diff
@@ -0,0 +1,276 @@
+// Implementation of a Merkle tree in Move. Supports constructing a new tree
+// with a given depth, as well as proving that a leaf node belongs to the tree.
+module pyth::merkle {
+    use std::vector::{Self};
+    use sui::hash::{keccak256};
+    use wormhole::bytes::{Self};
+    use wormhole::bytes20::{Self, Bytes20, data};
+    use wormhole::cursor::{Self, Cursor};
+    use pyth::deserialize::{Self};
+
+    const MERKLE_LEAF_PREFIX: u8 = 0;
+    const MERKLE_NODE_PREFIX: u8 = 1;
+    const MERKLE_EMPTY_LEAF_PREFIX: u8 = 2;
+
+    const E_DEPTH_NOT_LARGE_ENOUGH_FOR_MESSAGES: u64 = 1212121;
+
+    // take keccak256 of input data, then return 20 rightmost bytes of result
+    fun hash(bytes: &vector<u8>): Bytes20 {
+        let hashed_bytes = keccak256(bytes); // 32 byte hash
+        let cursor = cursor::new(hashed_bytes);
+        let _ = bytes::take_bytes(&mut cursor, 12);
+        let bytes20 = bytes::take_bytes(&mut cursor, 20);
+        cursor::destroy_empty(cursor);
+        bytes20::from_bytes(bytes20)
+    }
+
+    fun emptyLeafHash(): Bytes20 {
+        let v = vector::empty();
+        bytes::push_u8(&mut v, MERKLE_EMPTY_LEAF_PREFIX);
+        hash(&v)
+    }
+
+    fun leafHash(data: &vector<u8>): Bytes20 {
+        bytes::push_u8(&mut *data, MERKLE_LEAF_PREFIX);
+        hash(data)
+    }
+
+    fun nodeHash(
+        childA: Bytes20,
+        childB: Bytes20
+    ): Bytes20 {
+        if (greaterThan(childA, childB)) {
+            (childA, childB) = (childB, childA);
+        };
+        let data_A = bytes20::data(&childA);
+        let data_B = bytes20::data(&childB);
+        vector::append(&mut data_A, data_B);
+        bytes::push_u8(&mut data_A, MERKLE_NODE_PREFIX);
+        hash(&data_A)
+    }
+
+    // greaterThan returns whether a is strictly greater than b
+    // note that data(&a) and data(&b) are both vectors of length 20
+    fun greaterThan(a: Bytes20, b: Bytes20): bool{
+        // aa and bb both have length 20
+        let aa = data(&a);
+        let bb = data(&b);
+        let i = 0;
+        while (i < 20){
+            if (*vector::borrow(&aa, i) > *vector::borrow(&bb, i)){
+                return true
+            } else if (*vector::borrow(&bb, i) > *vector::borrow(&aa, i)){
+                return false
+            };
+            i = i + 1;
+        };
+        false
+    }
+
+    // The Sui Move stdlb insert function shifts v[i] and subsequent elements to the right.
+    // We don't want this behavior, so we define our own vector insert function using swapping.
+    // Reference: https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/packages/move-stdlib/sources/vector.move
+    fun vectorInsert<T: drop>(a: &mut vector<T>, value: T, index: u64){
```

Comment:
this is really setting the `i`th element to value. I would call this like `setElement`, as "insert" usually means something that does the shifting behavior above.